### ISSUE TITLE
Best way to do color?

### ISSUE THREAD

**Issue by 1sra3l** [MoAlyousef/fltk-form#5]
*Before* I do anything, I want your opinions on implementation.
The items needed to do this are:
* `let my_color:Color = fltk::enums::Color::from_html("#ffffff");` Similar to the [`from_rgb()`](https://docs.rs/fltk/1.2.17/fltk/enums/struct.Color.html#method.from_rgb) method
* `callback()` is [color popup](https://docs.rs/fltk/1.2.17/fltk/dialog/fn.color_chooser.html) that [replaces current widget color](https://docs.rs/fltk/1.2.17/fltk/prelude/trait.WidgetExt.html#tymethod.set_color).  Are callbacks possible with this? Especially being able to do [`set_label(string_html_value_returned_from_static_function)`](https://docs.rs/fltk/1.2.17/fltk/prelude/trait.WidgetExt.html#tymethod.set_label)
* pub FlColor(pub String);  // Same convention

If it is simple to make callbacks function here, I would like to also make other common dialogs, like `FlFile(String)` to open a file chooser.
It would also simply `set_label()` as well.  Perhaps a checkbox button with `fl_ask()` Using something like `FlAsk(pub value:bool, pub label:String);`
Can a callback set the struct fields?
```rs
// from the docs
but.set_callback(move |_| frame.set_label("Hello World!")); // the closure capture is mutable borrow to our button
```

**Comment by MoAlyousef** [comment]
Callbacks are possible, however you should not use set_raw_user_data() since it would override the callback. Currently the struct implementing FltkForm can't be modified from the callback. The file dialog thing should probably be separated from the form. Maybe a standalone button spawns the file dialog, then using set_prop, the file path can be modified.

**Comment by 1sra3l** [comment]
This is the function to create the html color string from FLTK color
```rs
pub fn convert_rgb_to_hex(r:u8, g:u8, b:u8) -> String {
    let mut hexa_code = String::from("#");
// std::fmt::UpperHex is `X` lower would be `x`
    hexa_code.push_str(format!("{:X}", r).as_str());
    hexa_code.push_str(format!("{:X}",g).as_str());
    hexa_code.push_str(format!("{:X}",b).as_str());
    hexa_code
}
```
I researched solutions and found [hex](https://crates.io/crates/hex) for decoding.  Do you have any better solution?

**Comment by MoAlyousef** [comment]
You can also get rgb values from an FLTK color using Color::to_rgb();
You can also use `format!("#{:X}{:X}{:X}", r, g, b)`

**Comment by 1sra3l** [comment]
@MoAlyousef good point, my C++ tendencies have not gone away it seems. Your code is much more concise and easier to read.
I was thinking the same thing about `to_rgb()` Perhaps a second wrapper function, so both can be available for people doing math?
What do you think about using that crate hex to convert a string in some sort of `Color::from_html(color:String)`  where it uses hex to `decode()`.  Would you rather not bring in another crate?  Or do you have a different solution?

I would like the html color stuff to be available in `fltk::enums::Color` so I want to know your take before I code it and make a PR.

**Comment by MoAlyousef** [comment]
I honestly don't see an advantage of Color::from_html("#fffaff") over Color::from_hex(0xfffaff). It would require extra checks for String sanity would have to return a `Result<Color, FltkError>`, it also can't be a const function. Also the hex crate seems to basically hash the string then encode it to hex. So I don't thing #ffffff would result in a correct color value.

**Comment by 1sra3l** [comment]
Understood about the hex crate. 

The advantage for `Color::from_html("#000000")` is for the color picker.  I am sure you use HTML colors in most of your system color pickers (text editor, graphics editor, etc..).  I think displaying/translating the color for the end-user is more in line with your level of professionalism in these libraries.  Most system level color pickers support html colors.  It may be advantageous to add this in the theme capability.
HTML colors are also pretty much the standard way to store color data in configuration files these days.

If you are still against the idea, I will be fine with whatever you decide.  You have a good vision for this project and I respect your scope of knowledge as being much more complete here, than mine.

**Comment by MoAlyousef** [comment]
I have added to fltk-rs Color::from_html_color and to_html_color:
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let white = Color::from_html_color("#ffffff").unwrap();
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    let mut btn = button::Button::new(160, 200, 80, 30, "Click");
    btn.set_color(white);
    dbg!(btn.color().to_html_color());
    win.end();
    win.show();
    a.run().unwrap();
}
```

**Comment by MoAlyousef** [comment]
I might change the name to hex_string or hex_str. The problem is that html color might be confused to refer to html color names:
https://www.w3schools.com/tags/ref_colornames.asp

**Comment by 1sra3l** [comment]
You know what, I had been looking in `fltk-theme` and forgot that the [colors](https://github.com/fltk-rs/fltk-theme/blob/main/src/colors/html.rs) were not in the main fltk.  
Please do change the name, it will be less confusing!
```rs
/// This function returns enums::Color from a string html color hexadecimal number
/// let rebecca_purple = enums::Color::from_hex_str("#663399");
/// assert_eq!(rebecca_purple.to_hex_str(), "#663399");
```
That makes sense.  It also looks nicer than `{to_,from_}html_color`
That will likely also make themes easier for people to port (from GTK/Qt), or create from specific color schemes.
Thanks for making such a small solution that does so much!

This will allow something like
```rs
FlColor("#663399")
```
to become a color chooser button, or for someone to read a configuration file into an interface.

I think I will close this, as I think all the things I listed as things to decide prior to starting have been done.

**Comment by 1sra3l** [comment]
@MoAlyousef 
Here is the function I wrote for the button.
```rs
pub fn color_button(color:&str) -> button::Button {
    let val = enums::Color::from_hex_str(color);
    let mut b = button::Button::default();

    let rgb = match val {
        Ok(rgb) => rgb.to_rgb(),
        Err(_e) => (102,51,153),
    };
    b.set_color(enums::Color::from_rgb(rgb.0, rgb.1, rgb.2));
    b.set_callback(|this| {
        let color = this.color();
        let label = enums::Color::to_hex_str(&color);
        let c = dialog::color_chooser_with_default(label.as_str(), dialog::ColorMode::Hex, color.to_rgb());
        let color = enums::Color::from_rgb(c.0, c.1, c.2);
        this.set_color(color);
    });
    b
}
```
I cannot make a real PR until `to_hex_str()` is in fltk (non git)  But you could just add this function yourself if you like.
I was thinking about also using:
```rs
/// Hexadecimal color string
/// ```
/// use fltk_form::FlHexaColor;
/// let rebecca_purple = FlHexaColor(String::from("#663399"));
/// ```
pub struct FlHexaColor(pub String);
impl fmt::Display for FlHexaColor {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}
```

**Comment by MoAlyousef** [comment]
Looks good. 1.2.20 should be released in 2 days. By the way, the colors that are used in fltk-theme are from the [color-maps crate](https://crates.io/crates/color-maps).
I don't think it should be added as dependency here, however, maybe it could be added as a dev-dependency and it can be used in an example. An adaptor function would be needed, something like Color::from_html_color(&str).

**Comment by 1sra3l** [comment]
Nice work!
Feel free to just put that function in however you see fit.
Looking at the line numbers in your source it looks like you have all 140 colors, though why are they not in alphabetical order?
How did you generate it, or did you type it?

Thanks for helping me make this!

One thing I think should change in fltk-form is
```rs
pub fn make_image_frame<P: AsRef<Path>>(filename: P) -> frame::Frame {
    let mut frame = frame::Frame::default(); // this set the label, long filenames might get messy
    frame.set_tooltip(filename.as_ref().to_str().unwrap()); // tooltip is a window and only shown when needed
//more code
}
```

**Comment by MoAlyousef** [comment]
I've applied the change to the image frame, now the path shows in the tooltip. I've also added a utils::color_map_adaptor, so the color_button function can now take an html name as well:
```rust
pub fn color_button(color:&str) -> button::Button {
    let val = enums::Color::from_hex_str(&utils::color_map_adaptor(color));
// rest of code
```

P.S. Regarding the 140 colors, I copied the values, then with block modification, turned it into a Rust HashMap :)

**Comment by 1sra3l** [comment]
@MoAlyousef wow, I think that will be really nice for people!  I know I will use it, even easier to use color names!

What do you mean by "block modification", I am unfamiliar with the term, though I looked at the code and saw the hash map.
I have written a script to read some text input file and then output code in whatever language I was using, is this similar to what you mean?

**Comment by MoAlyousef** [comment]
Visual studio code allows block/multiline editing which is a nice productivity boost, since you write code in a single line and it applies to all chosen 140 lines.

**Comment by MoAlyousef** [comment]
I've published fltk 1.2.20

**Comment by 1sra3l** [comment]
Thanks!
I guess this should be closed now
